### ISSUE TITLE
[explicit-module-boundary-types] [explicit-function-return-type] both rules result in duplicate error messages.

### ISSUE THREAD

**Issue by trusktr** [typescript-eslint/typescript-eslint#2719]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": "error",
    "@typescript-eslint/explicit-module-boundary-types": "error"
  }
}
```

```TS
export class Foo {
  test() {} // no return type
}
```

**Expected Result**

The error should be de-duplicated:

```
/home/trusktr/src/project/src/renderer/Foo.ts
  21:3  error    Missing return type on function  (@typescript-eslint/explicit-function-return-type, @typescript-eslint/explicit-module-boundary-types)
```

**Actual Result**

There are two of effectively the same error:

```
/home/trusktr/src/project/src/renderer/Foo.ts
  21:3  error    Missing return type on function  @typescript-eslint/explicit-module-boundary-types
  21:3  error    Missing return type on function  @typescript-eslint/explicit-function-return-type
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.6.0` |
| `@typescript-eslint/parser`        | `4.6.0` |
| `TypeScript`                       | `4.0` |
| `ESLint`                           | `7.12.1` |
| `node`                             | `14.8` |

**Comment by bradzacher** [comment]
Yes - they do.
They're not intended to be used together.

`explicit-module-boundary-types` is intended to be a mostly weaker version of `explicit-function-return-type`.
It did evolve with a few extra checks (specifically ensuring you type parameters etc), but it still only checks your module boundaries.

**Comment by trusktr** [comment]
@bradzacher Thanks! Which config would you choose regarding these two?

**Comment by bradzacher** [comment]
> Maybe a ESLint feature could fold same errors into one

That would be a request for the eslint repo! Plugins have no control over that - every rule is independent.

It's personal preference. Some people prefer to strictly type all function returns.
Some people think that's too much and only want to strictly type the boundaries.

It depends on what your coding style is.

I personally tend to annotate every function because I like to create clear and strict contracts.

**Comment by trusktr** [comment]
> tend to annotate every function because I like to create clear and strict contracts.

That's what I'd like too. Seems then the best way is to use `explicit-function-return-type` along with the `noImplicitAny: true` option in tsconfig (for function parameters), so that it covers everything besides the module boundaries. How does that sound?

**Comment by bradzacher** [comment]
I would never ***ever*** use TS without `noImplicitAny` turned on.
Doing so is a recipe for weakly typed broken code!

Personally I'd just turn on `strict` to ensure your code is as safe as possible.

**Comment by trusktr** [comment]
Of course, I agree with that. But I'm wondering, if `noImplicitAny` is turned on (or `strict`), then the ESLint rule is redundant right? Or is there another reason to have the lint rule too?

**Comment by bradzacher** [comment]
`noImplicitAny` just ensures you have type annotations when TS cannot infer the type.

The lint rules are there to help ensure that you're clearly defining strict contracts for your code.

They don't really add any safety to your codebase, but they do make it easier to read code in places like github without intellisense (and IMO they make it easier to read the code in general).

**Comment by trusktr** [comment]
Thanks for the insight @bradzacher, much appreciated. Makes sense about the readability, and something I aim for too.

It seems that with enforced type checking, it shouldn't be a problem (f.e. commit hooks and CI checks for PRs). With those enforcements, I _think_ that coupling the `no-explicit-any` and `explicit-function-return-type` lint rules with `noImplicitAny` (or `strict`) in tsconfig may do the trick, unless there's something I overlooked.

**Comment by bradzacher** [comment]
Yup all that help prevent common pitfalls.

If you're looking for an extra layer of security in case `any`s leak in from external libraries, then consider the `no-unsafe-*` set of lint rules. They help protect against accidentally using `any`s. They require type-aware linting but IMO are worth the price of admission!
